Turborepo Installation Fails on Windows - Missing Binaries

### Verify canary release

- [x] I verified that the issue exists in the latest Turborepo canary release.

### Link to code that reproduces this issue

s

### Which canary version will you have in your reproduction?

2.5.4

### Environment information

```block
### Environment Details
- **OS**: Windows_NT 10.0.26100 (Windows 11)
- **Node.js**: v20.17.0
- **npm**: v10.8.2
- **TurboJS Version**: 2.5.4
- **Shell**: PowerShell
- **Architecture**: x64 (from error context)
```

### Expected behavior

TruboJS should work on windows if installed 


### Actual behavior


### Issue Description
TurboJS installation fails on native Windows with missing binary errors, despite successful installation via npm. The same setup works correctly under WSL 2.0, suggesting a Windows-specific platform compatibility issue.

### Reproduction Steps
1. Fresh project setup on Windows
2. Run `npm install turbo --save-dev`
3. Attempt to run `turbo build`

### Observed Behavior
During installation:
```log
npm verbose reify failed optional dependency C:\Projects\turbo2\my-turborepo\node_modules\turbo-windows-arm64
npm verbose reify failed optional dependency C:\Projects\turbo2\my-turborepo\node_modules\turbo-windows-64
[similar errors for linux/darwin platforms]
```

When running Turbo:
```log
Turborepo did not find the correct binary for your platform (windows 64).
We were not able to find the binary at:
turbo-windows-64/bin/turbo.exe

Found unnecessary binaries for other platforms:
turbo-linux-64/bin/turbo
```

### Expected Behavior
1. TurboJS should install the correct Windows binary (`turbo-windows-64`)
2. `turbo build` should execute successfully without platform detection errors

### Troubleshooting Attempted
1. Clean reinstallation:
   ```powershell
   rd /s /q node_modules
   del package-lock.json
   npm cache clean --force
   npm install turbo --save-dev
   ```
2. Verified working in WSL 2.0 with identical versions
3. Checked all searched paths (listed in error) for binary presence



### To Reproduce

### Additional Observations
- The error suggests Turbo is looking for but not finding Windows binaries, while finding Linux binaries instead
- Works perfectly in WSL 2.0 environment with same Node/npm versions
- No Windows-specific installation caveats found in TurboJS documentation

### Potential Root Causes
1. Windows binary not being properly published/downloaded in npm package
2. Platform detection issue in TurboJS's postinstall script
3. PATH/environment variable interference on Windows
4. Permission issues during binary installation

### Support Questions
1. Is Windows fully supported for TurboJS installations?
2. Are there known issues with Windows 11 24H2 (build 26100)?
3. Are there required Windows dependencies (like VC++ redistributable)?

### Additional Helpful Information
- Does occur in both cmd.exe and PowerShell
- No antivirus interference detected
- User account has admin privileges
- Tried with both npm and yarn (same result)

### Additional context

_No response_

FWIW I've recently (since 6 days ago) started getting the same issue but on Mac (Intel) with the `turbo-darwin-64` binary and PNPM 9.x:

```
| Turborepo did not find the correct binary for your platform.
│ We will attempt to install it now.
│ Installation has failed.
│ ***
│ Turborepo failed to start.
│ Turborepo detected that you are running:
│ darwin 64
│ ***
│ We were not able to find the binary at:
│ turbo-darwin-64/bin/turbo
```

...and also:

```
 ERR_PNPM_ENOENT  ENOENT: no such file or directory, copyfile '/Users/{me}/Library/pnpm/store/v3/files/ad/fdf7211f7aeae93060a6fc9c967c7eaec34bfa782bf18d8bfca3048034b9593b4d2454e821f614c69c659122fe6b55d6d0dc15d875bd5ef23e3b82aa96ca2d-exec' -> '/Users/{me}/Documents/GitHub/{repo}/node_modules/.pnpm/turbo-darwin-64@2.5.5/node_modules/turbo-darwin-64_tmp_41044/bin/turbo'
```

It persisits even if I delete and re-clone the repo, or reinstall PNPM.   

I've also seen a recent report of a very similar issue but with `turbo-darwin-arm64` and PNPM 10.x - https://github.com/pnpm/pnpm/issues/9986 (8 days ago) - reported under PNPM but it looks very similar to this issue on NPM. 

So that's three similar issues within 10 days with three different binaries and both NPM and PNPM. Could be coincidental, could be related.

Similar issue from a year ago (closed as fixed in `turborepo@1.12.3`): https://github.com/vercel/turborepo/issues/7202

--- a/packages/turbo/bin/turbo
+++ b/packages/turbo/bin/turbo
@@ -36,6 +36,34 @@ function hasPackage(sourceObject, packageName) {
   return !!sourceObject[packageName] || !!sourceObject[`node_modules/${packageName}`];
 }
 
+// Binary packages were renamed from `turbo-<os>-<arch>` to `@turbo/<os>-<arch>`.
+// We try the scoped name first, then fall back to the legacy unscoped name so
+// that upgrades across the rename boundary still work without re-installing.
+function binaryPaths(platform, arch) {
+  const ext = platform === 'windows' ? '.exe' : '';
+  return [
+    `@turbo/${platform}-${arch}/bin/turbo${ext}`,
+    `turbo-${platform}-${arch}/bin/turbo${ext}`,
+  ];
+}
+
+function packageNames(platform, arch) {
+  return [
+    `@turbo/${platform}-${arch}`,
+    `turbo-${platform}-${arch}`,
+  ];
+}
+
+// Try require.resolve on each candidate, return the first that succeeds.
+function tryResolve(candidates) {
+  for (const candidate of candidates) {
+    try {
+      return require.resolve(candidate);
+    } catch (e) {}
+  }
+  return null;
+}
+
 // This provides logging messages as it progresses towards calculating the binary path.
 function getBinaryPath() {
   // First we see if the user has configured a particular binary path.
@@ -68,49 +96,53 @@ function getBinaryPath() {
 
   let { platform, arch } = process;
   // Node uses `win32` but we want to use `windows` for consistency with
-  // our package naming conventions (i.e. `turbo-windows-64`).
+  // our package naming conventions (i.e. `@turbo/windows-64`).
   if (platform === "win32") {
     platform = "windows";
   }
   const resolvedArch = arch === 'x64' ? '64' : arch;
-  const ext = platform === 'windows' ? '.exe' : '';
 
   // Try all places in order until we get a hit.
 
+  const isSupported = availablePlatforms.includes(platform) && availableArchs.includes(resolvedArch);
+
   // 1. The package which contains the binary we _should_ be running.
-  const correctBinary = availablePlatforms.includes(platform) && availableArchs.includes(resolvedArch) ? `@turbo/${platform}-${resolvedArch}/bin/turbo${ext}` : null;
-  if (correctBinary !== null) {
-    try {
-      return require.resolve(`${correctBinary}`);
-    } catch (e) {}
+  //    Try the scoped name (@turbo/<os>-<arch>) first, then the legacy name (turbo-<os>-<arch>).
+  const correctCandidates = isSupported ? binaryPaths(platform, resolvedArch) : [];
+  const resolved = tryResolve(correctCandidates);
+  if (resolved !== null) {
+    return resolved;
   }
 
   // 2. Install the binary that they need just in time.
-  if (SHOULD_INSTALL && correctBinary !== null) {
+  if (SHOULD_INSTALL && correctCandidates.length > 0) {
     console.warn('Turborepo did not find the correct binary for your platform.');
     console.warn('We will attempt to install it now.');
 
     try {
       installUsingNPM();
-      const resolvedPath = require.resolve(`${correctBinary}`);
-      console.warn('Installation has succeeded.');
-      return resolvedPath;
-    } catch (e) {
-      console.warn('Installation has failed.');
-    }
+      const resolvedPath = tryResolve(correctCandidates);
+      if (resolvedPath !== null) {
+        console.warn('Installation has succeeded.');
+        return resolvedPath;
+      }
+    } catch (e) {}
+    console.warn('Installation has failed.');
   }
 
   // 3. Both Windows and macOS ARM boxes can run x64 binaries. Attempt to run under emulation.
-  const alternateBinary = (arch === "arm64" && ['darwin', 'windows'].includes(platform)) ? `@turbo/${platform}-64/bin/turbo${ext}` : null;
-  if (SHOULD_ATTEMPT_EMULATED && alternateBinary !== null) {
-    try {
-      const resolvedPath = require.resolve(`${alternateBinary}`);
+  const alternateCandidates = (arch === "arm64" && ['darwin', 'windows'].includes(platform))
+    ? binaryPaths(platform, '64')
+    : [];
+  if (SHOULD_ATTEMPT_EMULATED && alternateCandidates.length > 0) {
+    const resolvedPath = tryResolve(alternateCandidates);
+    if (resolvedPath !== null) {
       console.warn(`Turborepo detected that you're running:\n${platform} ${resolvedArch}.`);
-      console.warn(`We were not able to find the binary at:\n${correctBinary}`);
-      console.warn(`We found a possibly-compatible binary at:\n${alternateBinary}`);
+      console.warn(`We were not able to find the binary at:\n${correctCandidates[0]}`);
+      console.warn(`We found a possibly-compatible binary at:\n${resolvedPath}`);
       console.warn(`We will attempt to run that binary.`);
       return resolvedPath;
-    } catch (e) {}
+    }
   }
 
   // We are not going to run `turbo` this invocation.
@@ -118,7 +150,7 @@ function getBinaryPath() {
 
   // Possible error scenarios:
   // - The user is on a platform/arch combination we do not support.
-  // - We somehow got detection wrong and never attempted to run the _actual_ `correctBinary` or `alternateBinary`.
+  // - We somehow got detection wrong and never attempted to run the _actual_ correct binary or alternate binary.
   // - The user doesn't have the correct packages installed for their platform.
 
   // Explain our detection attempt:
@@ -146,32 +178,33 @@ function getBinaryPath() {
     }
   }
 
-  if (correctBinary !== null) {
+  if (correctCandidates.length > 0) {
     console.error();
     console.error('***');
     console.error();
-    console.error(`We were not able to find the binary at:\n${correctBinary}`);
+    console.error(`We were not able to find the binary at:\n${correctCandidates.join('\nor: ')}`);
     console.error();
     console.error(`We looked for it at:`);
-    console.error(require.resolve.paths(correctBinary).join('\n'));
+    console.error(require.resolve.paths(correctCandidates[0]).join('\n'));
   }
-  if (alternateBinary !== null) {
+  if (alternateCandidates.length > 0) {
     console.error();
     console.error('***');
     console.error();
     console.error(`Your platform (${platform}) can sometimes run x86 under emulation.`);
-    console.error(`We did not find a possibly-compatible binary at:\n${alternateBinary}`);
+    console.error(`We did not find a possibly-compatible binary at:\n${alternateCandidates.join('\nor: ')}`);
     console.error();
     console.error(`We looked for it at:`);
-    console.error(require.resolve.paths(alternateBinary).join('\n'));
+    console.error(require.resolve.paths(alternateCandidates[0]).join('\n'));
   }
 
   // Investigate other failure modes.
 
   // Has the wrong platform's binaries available.
-  const availableBinaries = availablePlatforms.map(platform => availableArchs.map(arch => `@turbo/${platform}-${arch}/bin/turbo${platform === 'windows' ? '.exe' : ''}`)).flat();
-  const definitelyWrongBinaries = availableBinaries.filter(binary => binary !== correctBinary && binary !== alternateBinary);
-  const otherInstalled = definitelyWrongBinaries.filter(binaryPath => {
+  const ownCandidates = new Set([...correctCandidates, ...alternateCandidates]);
+  const allBinaries = availablePlatforms.flatMap(p => availableArchs.flatMap(a => binaryPaths(p, a)));
+  const wrongPlatformBinaries = allBinaries.filter(b => !ownCandidates.has(b));
+  const otherInstalled = wrongPlatformBinaries.filter(binaryPath => {
     try {
       return require.resolve(binaryPath);
     } catch (e) {}
@@ -195,7 +228,6 @@ function getBinaryPath() {
 
   // Check to see if we have partially-populated dependencies in the npm lockfile.
   const MAX_LOOKUPS = 10;
-  const availablePackages = availablePlatforms.map(platform => availableArchs.map(arch => `@turbo/${platform}-${arch}`)).flat();
 
   try {
     // Attempt to find project root.
@@ -213,9 +245,13 @@ function getBinaryPath() {
 
         // If we don't show up in the lockfile it's the wrong lockfile.
         if (hasPackage(sourceObject, 'turbo')) {
-          // Check to see if all of `turbo-<PLATFORM>-<ARCH>` is included.
-          const hasAllPackages = availablePackages.every(package => hasPackage(sourceObject, package));
-          if (!hasAllPackages) {
+          // A platform is covered if either the scoped or legacy package name is in the lockfile.
+          const hasAllPlatforms = availablePlatforms.every(p =>
+            availableArchs.every(a =>
+              packageNames(p, a).some(name => hasPackage(sourceObject, name))
+            )
+          );
+          if (!hasAllPlatforms) {
             console.error();
             console.error('***');
             console.error();

